lib/pf/Authentication/Source/TwitterSource.pm

```diff
@@ -0,0 +1,211 @@
+
+    my $ua = LWP::UserAgent->new;
+    my $req = HTTP::Request->new(POST => $request_token_url);
+    $req->content_type('application/x-www-form-urlencoded');
+    $req->content($qs);
+    my $response = $ua->request($req);
+
+    unless ($response->is_success) {
+      get_logger->("Couldn't execute request properly. Response is : ".$response->content);
+      return undef;
+    }
+
+    my $dummy_url = "http://localhost?".$response->content;
+
+    use URI;
+    my $url_object = new URI($dummy_url);
+
+    my %params = $url_object->query_form();
+
+    my $oauth_token = $params{'oauth_token'};
+    my $oauth_token_secret = $params{'oauth_token_secret'};
+
+    return $oauth_token;
+}
+
+=head2 get_access_token
+
+Get the access token through the Twitter API using the
+oauth_token + oauth_verifier
+
+This will also return the username of the user.
+
+=cut
+
+sub get_access_token {
+    my ($self, $oauth_token, $oauth_verifier) = @_;
+    my $access_token_url = $self->{protected_resource_url};
+
+    my $params = {
+      oauth_consumer_key => $self->{client_id},
+      oauth_nonce => time,
+      oauth_signature_method => "HMAC-SHA1",
+      oauth_timestamp => time,
+      oauth_version => "1.0",
+      oauth_token => $oauth_token,
+      oauth_verifier => $oauth_verifier,
+    };
+
+    my $qs = $self->simple_sign($access_token_url, $params);
+
+    my $ua = LWP::UserAgent->new;
+    my $req = HTTP::Request->new(POST => $access_token_url);
+    $req->content_type('application/x-www-form-urlencoded');
+    $req->content($qs);
+    my $response = $ua->request($req);
+
+    unless ($response->is_success) {
+      die "Couldn't execute request properly. Response is : ".$response->content;
+    }
+
+    my $dummy_url = "http://localhost?".$response->content;
+    my $url_object = new URI($dummy_url);
+    my %params = $url_object->query_form();
+
+    return {access_token => $params{'access_token'}, username => $params{'screen_name'}};
+}
+
+=head2 build_sorted_query
+
+Will sort the parameters in a hash and put them in a
+string reprensenting the parameters (param1=value1&param2=value2&)
+
+=cut
+
+sub build_sorted_query {
+  my ($self, $input) = @_;
+  my $qs;
+  foreach (sort keys %$input) {
+      $qs .= $_."=".$input->{$_}."&";
+  }
+  return substr ($qs, 0, -1);
+}
+
+=head2 simple_sign
+
+Will sign a query so it can be sent to the Twitter API
+See : https://dev.twitter.com/oauth/overview/creating-signatures
+
+=cut
+
+sub simple_sign {
+  my ($self, $url, $params) = @_;
+  my $IN = new CGI;
+
+  my $qs = $self->build_sorted_query($params);
+  my $signing_key = $IN->escape($self->{client_secret})."&";
+  my $signature_base = "POST&".$IN->escape($url)."&".$IN->escape($qs);
+
+  use Digest::HMAC_SHA1;
+  use MIME::Base64;
+
+  my $hmac = Digest::HMAC_SHA1->new($signing_key);
+  $hmac->add($signature_base);
+
+  $params->{oauth_signature} = $IN->escape(encode_base64($hmac->digest));
```

Comment:
I've tried it in c73c7038950e5bcbb260746a58c854831b46b53f and it doesn't sign properly so I'll leave the method that works